eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/AsyncPaymentTriggerer.scala

```diff
@@ -0,0 +1,142 @@
+  }
+
+  case class PeerPayments(notifier: ActorRef[PeerReadyNotifier.Command], pendingPayments: Set[Payment]) {
+    def update(currentBlockHeight: BlockHeight): Option[PeerPayments] = {
+      val expiredPayments = pendingPayments.filter(_.expired(currentBlockHeight))
+      expiredPayments.foreach(e => e.replyTo ! AsyncPaymentTimeout)
+      updatePaymentsOrStop(pendingPayments.removedAll(expiredPayments))
+    }
+
+    def cancel(paymentHash: ByteVector32): Option[PeerPayments] = {
+      val canceledPayment = pendingPayments.find(_.paymentHash == paymentHash)
+      if (canceledPayment.isDefined) canceledPayment.get.replyTo ! AsyncPaymentCanceled
+      updatePaymentsOrStop(pendingPayments.removedAll(canceledPayment))
+    }
+
+    private def updatePaymentsOrStop(pendingPayments: Set[Payment]): Option[PeerPayments] = {
+      if (pendingPayments.isEmpty) {
+        context.stop(notifier)
+        None
+      } else {
+        Some(PeerPayments(notifier, pendingPayments))
+      }
+    }
+
+    def trigger(): Unit = pendingPayments.foreach(e => e.replyTo ! AsyncPaymentTriggered)
+    def timeout(): Unit = pendingPayments.foreach(e => e.replyTo ! AsyncPaymentTimeout)
+  }
+
+  def start(): Behavior[Command] = {
+    context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[CurrentBlockHeight](WrappedCurrentBlockHeight))
+    watching(Map.empty)
+  }
+
+  private def watching(peers: Map[PublicKey, PeerPayments]): Behavior[Command] = {
+    Behaviors.receiveMessagePartial {
+      case Watch(replyTo, remoteNodeId, paymentHash, timeout) =>
+        peers.get(remoteNodeId) match {
+          case None =>
+            val notifier = context.spawn(
+              Behaviors.supervise(PeerReadyNotifier(remoteNodeId, switchboard, None)).onFailure(SupervisorStrategy.stop),
+              s"peer-ready-notifier-$remoteNodeId",
+            )
+            context.watchWith(notifier, NotifierStopped(remoteNodeId))
+            notifier ! NotifyWhenPeerReady(context.messageAdapter[PeerReadyNotifier.Result](WrappedPeerReadyResult))
+            val peer = PeerPayments(notifier, Set(Payment(replyTo, timeout, paymentHash)))
+            watching(peers + (remoteNodeId -> peer))
+          case Some(peer) =>
+            val peer1 = PeerPayments(peer.notifier, peer.pendingPayments + Payment(replyTo, timeout, paymentHash))
+            watching(peers + (remoteNodeId -> peer1))
+        }
+      case Cancel(paymentHash) =>
+        val peers1 = peers.flatMap {
+          case (remoteNodeId, peer) => peer.cancel(paymentHash).map(peer1 => remoteNodeId -> peer1)
+        }
+        watching(peers1)
+      case WrappedCurrentBlockHeight(CurrentBlockHeight(currentBlockHeight)) =>
+        val peers1 = peers.flatMap {
+          case (remoteNodeId, peer) => peer.update(currentBlockHeight).map(peer1 => remoteNodeId -> peer1)
+        }
+        watching(peers1)
+      case WrappedPeerReadyResult(PeerReadyNotifier.PeerReady(remoteNodeId, _)) =>
```

Comment:
You're not handling `PeerReadyNotifier.PeerUnavailable`! We should have a unit test for that, this is a bug.
It is safer to do:

```scala
case WrappedPeerReadyResult(result) => result match {
  ...
}
```

This will ensure that the compiler checks that all cases have been handled.

Reply:
Because we create `PeerReadyNotifier` with `timeout_opt=None`, it should never emit  `PeerReadyNotifier.PeerUnavailable`, and if it did we should ignore it because the `AsyncPaymentsTriggerer` watches for block height events that is uses to time out all watched peers. 

Do you think I should add a comment to the current code about why I do not handle `PeerReadyNotifier` ? or should I instead match the `PeerReadyNotifier` result and ignore it with a comment about why?

Reply:
> Do you think I should add a comment to the current code about why I do not handle PeerReadyNotifier ? or should I instead match the PeerReadyNotifier result and ignore it with a comment about why?

Yes, that would be better, because if we add a new `PeerReadyNotifier.Result` in the future, the compiler will let us know that we may need to handle it here, otherwise with the current code we won't know that we may be missing a case.

Reply:
Fixed and added a comment here: [9619b6e](https://github.com/ACINQ/eclair/pull/2491/commits/9619b6e3ee7c59cdfd93d3630dcce7a55d54c567)